src/player.cpp

```diff
@@ -10379,85 +10379,132 @@ hint_rating player::rate_action_takeoff( const item &it ) const
+                } );
+            if( iter == dependent.end() ) { // Not in the list yet
+                add_dependent( wit );
+                dependent.push_back( &wit );
+            }
+        }
+    };
+
+    if( is_worn( it ) ) {
+        add_dependent( it );
+    }
+
+    return dependent;
 }
 
-bool player::takeoff(int inventory_position, bool autodrop, std::vector<item> *items)
+bool player::takeoff( const item &it, std::function<bool(const item &)> interact )
 {
-    if (inventory_position == -1) {
-        return wield( ret_null );
+    auto iter = std::find_if( worn.begin(), worn.end(), [ &it ]( const item &wit ) {
+        return &it == &wit;
+    } );
+
+    if( iter == worn.end() ) {
+        add_msg_player_or_npc( m_info,
+            _( "You are not wearing that item." ),
+            _( "<npcname> is not wearing that item." ) );
+        return false;
+    }
+
+    for( const auto dep_it : get_dependent_worn_items( it ) ) {
+        if( !takeoff( *dep_it, interact ) ) {
+            return false; // Failed to takeoff a dependent item
+        }
     }
 
-    int worn_index = worn_position_to_index( inventory_position );
-    if( static_cast<size_t>( worn_index ) >= worn.size() ) {
-        add_msg_if_player( m_info, _("You are not wearing that item.") );
+    if( !interact( it ) ) {
         return false;
     }
-    bool taken_off = false;
 
-    auto first_iter = worn.begin();
-    std::advance( first_iter, worn_index );
-    item &w = *first_iter;
+    add_msg_player_or_npc( _( "You take off your %s." ),
+                           _( "<npcname> takes off their %s." ),
+                           it.tname().c_str() );
 
-    // Handle power armor.
-    if (w.is_power_armor() && w.covers(bp_torso)) {
-        // We're trying to take off power armor, but cannot do that if we have a power armor component on!
-        for( auto iter = worn.begin(); iter != worn.end(); ) {
-            item& other_armor = *iter;
+    mod_moves( -250 );    // TODO: Make this variable
+    iter->on_takeoff( *this );
+    worn.erase( iter );
 
-            if( &other_armor == &w || !other_armor.is_power_armor() ) {
-                ++iter;
-                continue;
-            }
-            if( !autodrop && items == nullptr ) {
-                add_msg_if_player( m_info, _("You can't take off power armor while wearing other power armor components.") );
-                return false;
+    recalc_sight_limits();
+    reset_encumbrance();
+
+    return true;
+}
+
+bool player::takeoff( const item &it )
+{
+    if( !get_dependent_worn_items( it ).empty() ) {
+        add_msg_player_or_npc( m_info,
+                               _( "You can't take off power armor while wearing other power armor components." ),
+                               _( "<npcname> can't take off power armor while wearing other power armor components." ) );
+        return false;
+    }
+
+    return takeoff( it, [ this ]( const item &it ) {
```

Comment:
Apart from the vector, that would also require a passed in boolean `interactive`. It's not the third argument I want to avoid, but the hard-coded interactivity. If we inject it into the function by means of a lambda, we can get whichever interactivity (e.g. various UI) / additional-filtering we want without polluting the actual logic. I'd prefer either to keep it, or implement another less hard-coded solution.